README.md
# Bash color

## Available colors

- BLACK
- RED
- GREEN
- BROWN
- BLUE
- PURPLE
- CYAN
- LIGHT_GRAY
- DEFAULT
- DARK_GRAY
- LIGHT_RED
- LIGHT_GREEN
- YELLOW
- LIGHT_BLUE
- LIGHT_PURPLE
- LIGHT_CYAN
- WHITE

## Available effect

- BOLD
- DIM (not working on Konsole)
- UNDERLINE
- BLINK (not working on Konsole and Gnome Terminal)
- INVERSE
- HIDDEN (not working on Konsole)

## Example

```python
from bashcolor import RED, UNDERLINE, colorize

print(colorize('Red color', RED))
print(colorize('Red background', background=RED))
print(colorize('Underline', effects=[UNDERLINE]))
```

## API

```python
def colorize(
    text,
    color=None,
    background=None,
    effects=[],
    color_256=None,
    background_256=None,
    with_end=True):
```


CHANGELOG.md
# Changelog

## [1.1.1](https://github.com/sbrunner/bashcolor/tree/1.1.1) (2021-08-09)

[Full Changelog](https://github.com/sbrunner/bashcolor/compare/1.1.0...1.1.1)

**Merged pull requests:**

- Add Prospector check [\#62](https://github.com/sbrunner/bashcolor/pull/62) ([sbrunner](https://github.com/sbrunner))

## [1.1.0](https://github.com/sbrunner/bashcolor/tree/1.1.0) (2021-08-01)

[Full Changelog](https://github.com/sbrunner/bashcolor/compare/1.0.1...1.1.0)

#### :package: Dependencies update

- Bump c2cciutils from 1.1.dev20210727151149 to 1.1.dev20210729075943 [\#61](https://github.com/sbrunner/bashcolor/pull/61) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210712112154 to 1.1.dev20210727151149 [\#60](https://github.com/sbrunner/bashcolor/pull/60) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210709152623 to 1.1.dev20210712112154 [\#59](https://github.com/sbrunner/bashcolor/pull/59) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210701074602 to 1.1.dev20210709152623 [\#58](https://github.com/sbrunner/bashcolor/pull/58) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210629154910 to 1.1.dev20210701074602 [\#57](https://github.com/sbrunner/bashcolor/pull/57) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210623154317 to 1.1.dev20210629154910 [\#56](https://github.com/sbrunner/bashcolor/pull/56) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210617081616 to 1.1.dev20210623154317 [\#55](https://github.com/sbrunner/bashcolor/pull/55) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210615102932 to 1.1.dev20210617081616 [\#54](https://github.com/sbrunner/bashcolor/pull/54) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210604063451 to 1.1.dev20210615102932 [\#53](https://github.com/sbrunner/bashcolor/pull/53) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210511092229 to 1.1.dev20210604063451 [\#51](https://github.com/sbrunner/bashcolor/pull/51) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210504062412 to 1.1.dev20210511092229 [\#50](https://github.com/sbrunner/bashcolor/pull/50) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210428132049 to 1.1.dev20210504062412 [\#49](https://github.com/sbrunner/bashcolor/pull/49) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210427101151 to 1.1.dev20210428132049 [\#48](https://github.com/sbrunner/bashcolor/pull/48) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210426091625 to 1.1.dev20210427101151 [\#47](https://github.com/sbrunner/bashcolor/pull/47) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210422085804 to 1.1.dev20210426091625 [\#46](https://github.com/sbrunner/bashcolor/pull/46) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210416145645 to 1.1.dev20210422085804 [\#45](https://github.com/sbrunner/bashcolor/pull/45) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump c2cciutils from 1.1.dev20210401100258 to 1.1.dev20210416145645 [\#43](https://github.com/sbrunner/bashcolor/pull/43) ([dependabot[bot]](https://github.com/apps/dependabot))
- Update Dependabot config file [\#40](https://github.com/sbrunner/bashcolor/pull/40) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.1.dev20210331152818 to 1.1.dev20210401100258 [\#39](https://github.com/sbrunner/bashcolor/pull/39) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.1.dev20210315182348 to 1.1.dev20210331152818 [\#38](https://github.com/sbrunner/bashcolor/pull/38) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.1.dev20210308113838 to 1.1.dev20210315182348 [\#37](https://github.com/sbrunner/bashcolor/pull/37) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.1.dev20210308105603 to 1.1.dev20210308113838 [\#36](https://github.com/sbrunner/bashcolor/pull/36) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210308083048 to 1.1.dev20210308105603 [\#35](https://github.com/sbrunner/bashcolor/pull/35) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210305083859 to 1.0.dev20210308083048 [\#32](https://github.com/sbrunner/bashcolor/pull/32) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210304083613 to 1.0.dev20210305083859 [\#31](https://github.com/sbrunner/bashcolor/pull/31) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210302092543 to 1.0.dev20210304083613 [\#30](https://github.com/sbrunner/bashcolor/pull/30) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210302075833 to 1.0.dev20210302092543 [\#29](https://github.com/sbrunner/bashcolor/pull/29) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210217133842 to 1.0.dev20210302075833 [\#28](https://github.com/sbrunner/bashcolor/pull/28) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump c2cciutils from 1.0.dev20210212123011 to 1.0.dev20210217133842 [\#27](https://github.com/sbrunner/bashcolor/pull/27) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

test.py
import bashcolor
from pyannotate_runtime import collect_types

if __name__ == '__main__':
    collect_types.init_types_collection()
    with collect_types.collect():

        bashcolor.print_colors()

        bashcolor.colorize("test", bashcolor.BLACK, bashcolor.RED, [bashcolor.BOLD])
        bashcolor.colorize("test", color_256=16, background_256=164)

    collect_types.dump_stats('type_info.json')


bashcolor/__init__.py
# see also: http://misc.flogisoft.com/bash/tip_colors_and_formatting


from typing import List, Optional

RESET = 0
BLACK = 30
RED = 31
GREEN = 32
BROWN = 33
BLUE = 34
PURPLE = 35
CYAN = 36
LIGHT_GRAY = 37
_OTHER = 38
DEFAULT = 39
DARK_GRAY = 90
LIGHT_RED = 91
LIGHT_GREEN = 92
YELLOW = 93
LIGHT_BLUE = 94
LIGHT_PURPLE = 95
LIGHT_CYAN = 96
WHITE = 97

BOLD = 1
DIM = 2  # not working on Konsole
UNDERLINE = 4
BLINK = 5  # not working on Konsole and gnome Terminal
INVERSE = 7
HIDDEN = 8  # not working on Konsole

_ESC = "\033"
_ESC = "\x1b"
_BACKGROUND = 10
_RESET_EFFECT = 20


def colorize(
    text: str,
    color: Optional[int] = None,
    background: Optional[int] = None,
    effects: Optional[List[int]] = None,
    color_256: Optional[int] = None,
    background_256: Optional[int] = None,
    with_end: bool = True,
) -> str:
    """
    Colorize a text.

    Arguments:

        text: the text to be colorize
        color: the text color code
        background: the background color code
        effects: the effect code (BOLD, DIM, UNDERLINE, BLINK, INVERSE, HIDDEN)
        color_256: The 256 text color code, use print_colors to get the possible values
        background_256: The 256 background color code, use print_colors to get the possible values
        with_end: Set to False to don't add the end tag
    """
    if effects is None:
        effects = []
    start = []
    end = []

    if color is not None:
        start.append(color)
        end.append(DEFAULT)
    elif color_256 is not None:
        start += [_OTHER, 5, color_256]
        end.append(DEFAULT)

    if background is not None:
        start.append(background + _BACKGROUND)
        end.append(DEFAULT + _BACKGROUND)
    elif background_256 is not None:
        start += [_OTHER + _BACKGROUND, 5, background_256]
        end.append(DEFAULT + _BACKGROUND)

    for effect in effects:
        start.append(effect)
        end.append(effect + _RESET_EFFECT)

    start_code = f"{_ESC}[{';'.join([str(s) for s in start])}m" if text != "" else ""
    end_code = f"{_ESC}[{';'.join([str(e) for e in end])}m" if with_end else ""
    return f"{start_code}{text}{end_code}"


def print_colors() -> None:
    """Print a small help with all the color."""
    color_pivot = [0]
    color_pivot += [e * 6 + 16 for e in range(37)]
    color_pivot.append(256)
    color_pivot_start = color_pivot[:-1]
    color_pivot_end = color_pivot[1:]
    color_table_list = [range(cs, ce) for cs, ce in zip(color_pivot_start, color_pivot_end)]

    for color_table in color_table_list:
        text = ""
        for color in color_table:
            color_string = str(color)
            padding = "".join([" " for e in range(3 - len(color_string))])
            text += colorize(f" {padding}{color_string} ", background_256=color, with_end=False)
        print(text + colorize("", background=DEFAULT))


if __name__ == "__main__":
    print_colors()
